Pulsar 2.9.3 release notes (#149)
apache/pulsar-site: site2/website-next/release-notes/versioned/pulsar-2.9.3.md

- [fix][broker] Fix REST produce msg redirect issue [#15551](https://github.com/apache/pulsar/pull/15551)
- [fix][broker] Fix call sync method in onPoliciesUpdate method [#13885](https://github.com/apache/pulsar/pull/13885)
- [fix][broker] Fix call sync method in onPoliciesUpdate method [#15227](https://github.com/apache/pulsar/pull/15227)
- [fix][broker] Fix cannot delete namespace with system topic [#14730](https://github.com/apache/pulsar/pull/14730)
- [fix][broker] Fix creating producer failure when set backlog quota [#15663](https://github.com/apache/pulsar/pull/15663)
- [fix][broker] Fix creating system namespace topic failure [#14949](https://github.com/apache/pulsar/pull/14949)
- [fix][broker] Fix deadlock in broker after race condition in topic creation failure [#15570](https://github.com/apache/pulsar/pull/15570)
- [fix][broker] Fix getPendingAckInternalStats redirect issue [#14876](https://github.com/apache/pulsar/pull/14876)
- [fix][broker] Fix inconsistent prompt message when schema version is empty using AVRO [#14626](https://github.com/apache/pulsar/pull/14626)

> - [fix][broker] Fix incorrect entryId in warn log when reading entry from tiered storage [#14685](https://github.com/apache/pulsar/pull/14685)

Reviewer: ```suggestion
- [fix][broker] Fix incorrect entryId in warning logs when reading an entry from tiered storage [#14685](https://github.com/apache/pulsar/pull/14685)
```

- [fix][broker] Fix precision issue and initial value for Consumer#avgMessagesPerEntry [#14666](https://github.com/apache/pulsar/pull/14666)
- [fix][broker] Fix problem at RateLimiter#tryAcquire [#15306](https://github.com/apache/pulsar/pull/15306)
- [fix][broker] Fix producerFuture not completed in ServerCnx#handleProducer [#14467](https://github.com/apache/pulsar/pull/14467)
- [fix][broker] Fix race condition between timeout and completion  in `OpAddEntry` [#15233](https://github.com/apache/pulsar/pull/15233)
- [fix][broker] Fix race condition in updating lastMarkDeleteEntry field [#15031](https://github.com/apache/pulsar/pull/15031)
- [fix][broker] Fix rewind failed when ``redeliverUnacknowledgedMessages`` [#15046](https://github.com/apache/pulsar/pull/15046)
- [fix][broker] Fix topic policy reader close bug [#14897](https://github.com/apache/pulsar/pull/14897)
- [fix][broker] Fix typo in enum name and handle closing of the channel properly since writeAndFlush is asynchronous [#15384](https://github.com/apache/pulsar/pull/15384)
- [fix][broker] Fix when nextValidLedger is null caused NPE [#13975](https://github.com/apache/pulsar/pull/13975)

> - [fix][broker] Fix wrong prompt exception when get non-persistent topic list without GET_BUDNLE permission [#14638](https://github.com/apache/pulsar/pull/14638)

Reviewer: ```suggestion
- [fix][broker] Fix wrong prompt exception when getting the non-persistent topic list without GET_BUDNLE permission [#14638](https://github.com/apache/pulsar/pull/14638)
```

- [fix][broker] Fix wrong state for non-durable cursor [#14869](https://github.com/apache/pulsar/pull/14869)

> - [fix][broker] Fix wrong unit of NIC speed on linux [#15770](https://github.com/apache/pulsar/pull/15770)

Reviewer: ```suggestion
- [fix][broker] Fix wrong unit of NIC speed on Linux [#15770](https://github.com/apache/pulsar/pull/15770)
```

- [fix][broker] Return if reset in progress [#14978](https://github.com/apache/pulsar/pull/14978)
- [fix][broker] Use dispatchRateLimiterLock to update dispatchRateLimiter [#15601](https://github.com/apache/pulsar/pull/15601)
- [fix][broker] When skipping updating mark delete position, execute callback with executor to prevent deadlock [#15971](https://github.com/apache/pulsar/pull/15971)
- [fix][broker] Expose configurationMetadataStore and localMetadataStore [#15661](https://github.com/apache/pulsar/pull/15661)
- [fix][broker] Filter the virtual NIC with relative path [#14829](https://github.com/apache/pulsar/pull/14829)
- [fix][broker] Fix MetadataStoreException$NotFoundException while doing topic lookup [#15633](https://github.com/apache/pulsar/pull/15633)
- [fix][broker] Fix calculation in getNumberOfEntriesInStorage [#15627](https://github.com/apache/pulsar/pull/15627)
- [fix][broker] Fix error log miss stack trace when create tenant fail [#14366](https://github.com/apache/pulsar/pull/14366)
- [fix][broker] Fix resource group does not report usage [#15292](https://github.com/apache/pulsar/pull/15292)

> - [fix][broker] Fix the pid occupied check when use pulsar-daemon start or stop process [#15745](https://github.com/apache/pulsar/pull/15745)

Reviewer: Is this item duplicate with #14701 in Line73?

Author: Oh, thanks! I will fix it ASAP.

Author: Fixed.

- [fix][broker] Fix duplicated delayed messages when all consumers disconnect [#14740](https://github.com/apache/pulsar/pull/14740)

> - [fix][broker] Fix the pid occupied check when use pulsar-daemon start or stop process [#14701](https://github.com/apache/pulsar/pull/14701)

Reviewer: ```suggestion
- [fix][broker] Fix the pid occupied check when using pulsar-daemon start or stop process [#14701](https://github.com/apache/pulsar/pull/14701)
```

- [fix][broker] Fix potential NPE in Replicator [#15003](https://github.com/apache/pulsar/pull/15003)

> - [improve][broker] Add log when update namespace policies with error [#14850](https://github.com/apache/pulsar/pull/14850)

Reviewer: ```suggestion
- [improve][broker] Add log when updating namespace policies with error [#14850](https://github.com/apache/pulsar/pull/14850)
```

- [improve][broker] Add publishRateLimitedTimes to topic metrics [#15739](https://github.com/apache/pulsar/pull/15739)
- [improve][broker] Avoid using blocking calls for the async method ``checkTopicOwnership`` [#15023](https://github.com/apache/pulsar/pull/15023)
- [improve][broker] Cancel offload tasks when managed ledger closed [#14545](https://github.com/apache/pulsar/pull/14545)
- [improve][broker] Close connection if a ping or pong message cannot be sent [#15382](https://github.com/apache/pulsar/pull/15382)

> - [improve][broker] Configure DLog Bookie, Pulsar, and Admin clients via pass through config [#15818](https://github.com/apache/pulsar/pull/15818)

Reviewer: ```suggestion
- [improve][broker] Configure DLog Bookie, Pulsar, and Admin clients via pass-through config [#15818](https://github.com/apache/pulsar/pull/15818)
```

> - [improve][broker] Full-support ssl provider, ciphers and protocols for broker service and proxy service [#14569](https://github.com/apache/pulsar/pull/14569)

Reviewer: ```suggestion
- [improve][broker] Full-support SSL provider, ciphers, and protocols for broker service and proxy service [#14569](https://github.com/apache/pulsar/pull/14569)
```

- [fix][txn] Fix some exception handle in transaction buffer [#14808](https://github.com/apache/pulsar/pull/14808)
- [fix][txn] Fix transaction buffer no snapshot close recover reader [#14830](https://github.com/apache/pulsar/pull/14830)
- [fix][txn] Fix transaction buffer recover BrokerMetadataException close topic [#14709](https://github.com/apache/pulsar/pull/14709)
- [fix][txn] Fix transaction buffer recover reader and writer fail [#14801](https://github.com/apache/pulsar/pull/14801)
- [fix][txn] Fix transaction buffer recover throw cursor already close [#14807](https://github.com/apache/pulsar/pull/14807)
- [fix][txn] Fix transaction log recover throw cursor already close [#14810](https://github.com/apache/pulsar/pull/14810)
- [fix][txn] Fix transaction pending ack store managed ledger WriteFail state [#14738](https://github.com/apache/pulsar/pull/14738)
- [fix][txn] Fix transaction producer stuck problem [#15061](https://github.com/apache/pulsar/pull/15061)
- [fix][txn] Fix transaction component recover fillQueue [#15418](https://github.com/apache/pulsar/pull/15418)
- [fix][txn] Fix transasction ack batch message [#15875](https://github.com/apache/pulsar/pull/15875)

> - [improve][txn] Avoid create multiple future and exception handler [#15089](https://github.com/apache/pulsar/pull/15089)

Reviewer: ```suggestion
- [improve][txn] Avoid creating multiple future and exception handlers [#15089](https://github.com/apache/pulsar/pull/15089)
```

- [improve][txn] Support configurable ``transactionBufferClientOperationTimeoutInMills`` [#15011](https://github.com/apache/pulsar/pull/15011)

#### Security

- [fix][auth] Add timeout of sync methods and  avoid call sync method for AuthoriationService [#15694](https://github.com/apache/pulsar/pull/15694)
- [fix][auth] Fix debug log authenticate role error [#14784](https://github.com/apache/pulsar/pull/14784)
- [fix][auth] Fix grant all permissions but can't list topic [#15501](https://github.com/apache/pulsar/pull/15501)
- [fix][auth] Fix handling single role and non-jwt-token in MultiRolesTokenAuthorizationProvider [#14857](https://github.com/apache/pulsar/pull/14857)
- [fix][auth] Generate correct well-known OpenID configuration URL [#15928](https://github.com/apache/pulsar/pull/15928)
- [fix][auth] Role with namespace produce authz can also get topics [#15740](https://github.com/apache/pulsar/pull/15740)
- [improve][auth] Add KeyStore support in WebSocket, Function Worker HTTPS Servers  [#15084](https://github.com/apache/pulsar/pull/15084)
- [improve][auth] Allow to config web server's cipher and protocols [#13354](https://github.com/apache/pulsar/pull/13354)

> - [improve][auth] Full-support set ssl provider, ciphers and protocols [#13740](https://github.com/apache/pulsar/pull/13740)

Reviewer: ```suggestion
- [improve][auth] Full-support set SSL provider, ciphers, and protocols [#13740](https://github.com/apache/pulsar/pull/13740)
```

- [improve][sql] Add Java version trim agent for presto 332 [#15236](https://github.com/apache/pulsar/pull/15236)
- [improve][sql] Pulsar SQL support for Decimal data type [#15153](https://github.com/apache/pulsar/pull/15153)

#### Function

- [fix][function] Pass configured metricsPort to k8s runtime [#14502](https://github.com/apache/pulsar/pull/14502)
- [fix][function] Check executor null when close the FileSource [#15247](https://github.com/apache/pulsar/pull/15247)
- [fix][function] Fix pulsar-managed runtimes failed start function with package URL from package management service [#14814](https://github.com/apache/pulsar/pull/14814)
- [fix][function] Fix some IOExceptions when create functions from package URL [#14553](https://github.com/apache/pulsar/pull/14553)
- [fix][function] Handle NPE when `getLeader` returns null [#15058](https://github.com/apache/pulsar/pull/15058)
- [fix][function] Provide default error handler for function log appender [#15728](https://github.com/apache/pulsar/pull/15728)
- [fix][function] Refine file io connector [#15250](https://github.com/apache/pulsar/pull/15250)

#### Connector

> - [fix][connector] Pass client builder if no service url provided to debezium connector [#12145](https://github.com/apache/pulsar/pull/12145)

Reviewer: ```suggestion
- [fix][connector] Pass client builder if no service URL is provided to Debezium connector [#12145](https://github.com/apache/pulsar/pull/12145)
```

> - [fix][connector] Throw exceptions when kafka offset backing store failed to start  [#14491](https://github.com/apache/pulsar/pull/14491)

Reviewer: ```suggestion
- [fix][connector] Throw exceptions when Kafka offset backing store failed to start  [#14491](https://github.com/apache/pulsar/pull/14491)
```

#### Tiered Storage

- [fix][tiered-storage] Fix NoClassDefFoundError: com/google/inject/AbstractModule in pulsar-io/batch-data-generator and Jcloud offloader [#14150](https://github.com/apache/pulsar/pull/14150)
- [improve][tiered-storage] Add debug information  [#14907](https://github.com/apache/pulsar/pull/14907)
- [improve][tiered-storage] Add pure S3 provider for the offloader [#15710](https://github.com/apache/pulsar/pull/15710)

> - [improve][tiered-storage] Reduce cpu usage when offloading the ledger [#15063](https://github.com/apache/pulsar/pull/15063)

Reviewer: ```suggestion
- [improve][tiered-storage] Reduce CPU usage when offloading the ledger [#15063](https://github.com/apache/pulsar/pull/15063)
```
